autopush/router/adm.py

```diff
@@ -0,0 +1,206 @@
+"""ADM Router"""
+import time
+import requests
+
+from typing import Any  # noqa
+
+from requests.exceptions import ConnectionError
+from twisted.internet.threads import deferToThread
+from twisted.logger import Logger
+
+from autopush.exceptions import RouterException
+from autopush.metrics import make_tags
+from autopush.router.interface import RouterResponse
+from autopush.types import JSONDict  # noqa
+
+
+class ADMAuthError(Exception):
+    pass
+
+
+class ADMClient(object):
+    def __init__(self,
+                 credentials=None,
+                 logger=None,
+                 metrics=None,
+                 endpoint="api.amazon.com",
+                 **options
+                 ):
+
+        self._client_id = credentials["client_id"]
+        self._client_secret = credentials["client_secret"]
+        self._token_exp = 0
+        self._auth_token = None
+        self._aws_host = endpoint
+        self._logger = logger
+        self._metrics = metrics
+        self._request = requests
+
+    def refresh_key(self):
+        url = "https://{}/auth/O2/token".format(self._aws_host)
+        if self._auth_token is None or self._token_exp < time.time():
+            body = dict(
+                grant_type="client_credentials",
+                scope="messaging:push",
+                client_id=self._client_id,
+                client_secret=self._client_secret
+            )
+            headers = {
+                "content-type": "application/x-www-form-urlencoded"
+            }
+            resp = self._request.post(url, data=body, headers=headers)
+            if resp.status_code != 200:
+                self._logger.error("Could not get ADM Auth token {}".format(
+                    resp.text
+                ))
+                raise ADMAuthError("Could not fetch auth token")
+            reply = resp.json()
+            self._auth_token = reply['access_token']
+            self._token_exp = time.time() + reply.get('expires_in', 0)
+
+    def send(self, reg_id, payload, ttl=None, collapseKey=None):
+        self.refresh_key()
+        headers = {
+            "Authorization": "Bearer {}".format(self._auth_token),
+            "Content-Type": "application/json",
+            "X-Amzn-Type-Version":
+                "com.amazon.device.messaging.ADMMessage@1.0",
+            "X-Amzn-Accept-Type":
+                "com.amazon.device.messaging.ADMSendResult@1.0",
+            "Accept": "application/json",
+        }
+        data = {}
+        if ttl:
+            data["expiresAfter"] = ttl
+        if collapseKey:
+            data["consolidationKey"] = collapseKey
+        data["data"] = payload
+        url = ("https://api.amazon.com/messaging/registrations"
+               "/{}/messages".format(reg_id))
+        resp = self._request.post(
```

Comment:
same timeout here

Reply:
We can't use an async HTTP request? I know some of the router calls use third party tooling that we couldn't use the twisted http client with, but we should be able to here, right?

Reply:
I'd prefer to do the async requests as a separate issue, since that will span multiple files and fuzz exactly what this patch is supposed to address.